### PR TITLE
[DateLoader] more clearly expose 'default_collate' and 'default_convert' to users

### LINKED ISSUE #69445: [Feature Request] Expose default_collate as a public method
## 🚀 Feature

Currently, `default_collate` resides in the private module `torch.utils.data._utils.collate `. It would be nice to expose `default_collate` as a public method, in the case one wants to write a custom `collate_fn` that wraps the default one.

## Motivation

Sometimes one would like to perform some default processing to each batch, but still leverage the default `collate_fn` behaviour to do all the lifting of type conversion etc.

## Pitch

Make `default_collate_fn` a public method and document its behaviour in https://pytorch.org/docs/stable/data.html

## Alternatives

One could possibly introduce another attribute `post_processor` that applies another `Callable` after `collate_fn`. 

cc @SsnL @VitalyFedyunin @ejguan @NivekT

### PR DESCRIPTION
**Pull request by NivekT**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #69862

Fixes #69445

cc @SsnL @VitalyFedyunin @ejguan @NivekT

Differential Revision: [D33068792](https://our.internmc.facebook.com/intern/diff/D33068792)

### CODE REVIEW

**Review by ejguan — APPROVED** [review]

**Comment by SsnL** [comment]
Can we not expose functions with no complete docstrings?

**Comment by SsnL** [comment]
Also I would much rather have a complete solution (e.g. https://github.com/pytorch/pytorch/issues/33181) than small patches like this. Maybe DataPipe has something for this, but I don't know the details.

**Comment by NivekT** [comment]
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33068792).

**Comment by ejguan** [comment]
> Can we not expose functions with no complete docstrings?

Agree! It should be easily achieved https://github.com/pytorch/pytorch/pull/69851#issuecomment-992694879

> Also I would much rather have a complete solution (e.g. #33181) than small patches like this. Maybe DataPipe has something for this, but I don't know the details.

I don't think we would do that much effort because DataPipe would provide the flexibility for users to use custom collate as a callable function and apply it on different datasets.

**Comment by NivekT** [comment]
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33068792).

File: torch/utils/data/_utils/collate.py [hunk 1]
```diff
@@ -51,8 +82,49 @@ def default_convert(data):
 
 
 def default_collate(batch):
-    r"""Puts each data field into a tensor with outer dimension batch size"""
+    r"""
```
Comment (NivekT) [inline torch/utils/data/_utils/collate.py:85]: @ejguan Have a look and let me know what you think.

File: torch/utils/data/_utils/collate.py [hunk 2]
```diff
@@ -51,8 +82,49 @@ def default_convert(data):
...
+        Here is the general input type (based on the type of the element within the batch) to output type mapping:
+        * :class:`torch.Tensor` -> :class:`torch.Tensor` (with an added outer dimension batch size)
+        * NumPy Arrays -> :class:`torch.Tensor`
+        * `float` -> :class:`torch.Tensor`
+        * `int` -> :class:`torch.Tensor`
+        * `str` -> `str` (unchanged)
+        * `bytes` -> `bytes` (unchanged)
+        * `Mapping[K, V]` -> `Mapping[K, default_collate(V)]`
+        * `NamedTuple[V]` -> `NamedTuple[default_collate(V)]`
+        * `Sequence[V]` -> `Sequence[default_collate(V)]`
```
Comment (ejguan) [inline torch/utils/data/_utils/collate.py:102]: `default_collate(V)` can make users confused as single value from a batch. Could we do:
`Mapper[K, V_i]` -> `Mapping[K, default_collate([V_1, V_2, ...])]`
`NamedTuple[V1_i, V2_i]` -> `NamedTuple[default_collate([V1_1, V1_2, ...]), default_collate([V2_1, V2_2, ...])]`
`Sequence[V1_i, V2_i, ...]` -> `Sequence[default_collate([V1_1, V1_2, ...]), default_collate([V2_1, V2_2, ...]), ...]`

**Review by ejguan — APPROVED** [review]
LGTM with a nit comment.

**Comment by NivekT** [comment]
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33068792).

**Pull request closed without merging**